transactions/service.go

```diff
@@ -18,37 +19,68 @@ type Service interface {
 	SaveTransaction(tx *Transaction) error
 	GetTransactionStatus(identity identity.CentID, id uuid.UUID) (*transactionspb.TransactionStatusResponse, error)
 	WaitForTransaction(accountID identity.CentID, txID uuid.UUID) error
+	RegisterHandler(txID uuid.UUID, onUpdate func(status Status) error)
 }
 
 // NewService returns a Service implementation.
 func NewService(repo Repository) Service {
-	return service{repo: repo}
+	return &service{repo: repo, handlers: make(map[uuid.UUID][]func(status Status) error)}
 }
 
 // service implements Service.
 type service struct {
-	repo Repository
+	repo     Repository
+	handlers map[uuid.UUID][]func(status Status) error
+	mu       sync.Mutex
 }
 
 // SaveTransaction saves the transaction.
-func (s service) SaveTransaction(tx *Transaction) error {
-	return s.repo.Save(tx)
+func (s *service) SaveTransaction(tx *Transaction) error {
+	err := s.repo.Save(tx)
+	if err != nil {
+		return err
+	}
+
+	if len(tx.Logs) < 1 {
+		return nil
+	}
+
+	s.mu.Lock()
+
+	for _, h := range s.handlers[tx.ID] {
```

Comment:
Maybe we can move the call of the handlers in an own function together with a comment?

Reply:
Is it correct that multiple queuing task are using the same tx.ID?

If that is the case it might cause a problem here:

Let's say task A registers a handler to trigger task B.
B has an own handler as well.

After finishing task A this loop will be executed.
The handler for A would register the handler for task B and queue the task.

Now we have two entries in this list. 
Could it be that the second handler func is now executed before task B is finished?
(Because we are still in the loop)

Reply:
Also this handler lock is kind of scary,
as this can lead to deadlocks very easily since as someone might create a dependancy between a downstream task and an upstream handler and since handler is arbitrary code it can hold the lock indefinitely I think current solution is very risky in the long term. I have doubts whether we should fix it now when it's easier to fix in this pr.

Reply:
The way I would approach this is:
- Enable nested transactions out of a parent tx
- Create a tx for each task that is done async.
- Pass through the context the parent tx
- Update the parent tx with every update of the child ones
- The only place where a new child tx is created is by the previous task service (no handlers, or callbacks)
- Only the last task should update the latest status of the parent tx

Let's discuss this tomorrow

Reply:
@vedhavyas 
A global lock that serialises all transactions not even related to each other is a problem in any system. Since that means a bug or a problems caused by a single handler can cause all other non related transactions to block. You probably have heard about python GIL, its a similar situation here.

I agree that concern that @manuelpolzhofer is could indeed be rare, and thats not the reason I don't agree with this change as you may observe.